### ISSUE TITLE
[BUG] Parser no longer works correctly after interruption in the middle of big packet

### ISSUE THREAD

**Issue by serfreeman1337** [bertmelis/espMqttClient#34]
**Describe the bug**
Device lost connection during OTA update and after that it stuck in reconnection loop.
OTA update performs in the chunks of qos1 publish packets with length of 65536 bytes.

Logs:
```
Disconnecting, server exceeded keepalive
:wustmo
:close
clearing queue (clear session: false)
premature end: res:0, pos:65536/286395 <-- debug log from OTA
---> mqtt.connect()
payload: index 1582, total 65536, avail 4/4
Disconnecting, expected CONNACK - protocol error
:close
clearing queue (clear session: false)
---> mqtt.connect()
payload: index 1586, total 65536, avail 4/4
Disconnecting, expected CONNACK - protocol error
:close
clearing queue (clear session: false)
```

**Comment by bertmelis** [comment]
There are a few things wrong that, when happen at the same time, cause the observed behaviour.

- when nearing keep alive the client sends a ping but the server doesn't respond because it's I nthe middle of sending a large message.
- when reconnecting, the parser is still parsing the big message whereas it should parse connack
- this probably repeats be ause on qos1, the message is never acknowledged.

It basically comes down to a scenario where receiving a message takes longer then the keepalive time. It might also be a problem when sending takes longer then keepalive.

This has to be solved of course.

**Comment by bertmelis** [comment]
Untill this is fixed, I think you can 'evade' this issue by raising the keepalive value.

Unless somebody comes up with a solution I will need a couple of days to think about this.

**Comment by bertmelis** [comment]
Do you also have logs for just before the disconnect?
The client should not disconnect because of timeout in the middle of receiving.

**Comment by serfreeman1337** [comment]
Steps to reproduce:
1. Send PUBLISH with big payload (like 64k)
2. Block internet connection while device is still processing this publish
3. Wait for disconnect due to exceeded keep alive
4. Unblock internet connection

> Do you also have logs from just before the disconnect?

<details>
  <summary>
    debug
  </summary>

```
12:02:43.145 -> rx len 1440
12:02:43.145 -> Packet type: 0x40
12:02:43.145 -> Remaining length: 2
12:02:43.145 -> Packet complete
12:02:43.145 -> Parsed 4 - remaining 1436
12:02:43.145 -> Packet type: 0x32
12:02:43.145 -> Remaining length: 65564
12:02:43.145 -> Packet variable header topic complete
12:02:43.145 -> Packet variable header complete
12:02:43.145 -> payload: index 0, total 65536, avail 1404/1436
12:02:43.145 -> Parsed 1436 - remaining 0
12:02:43.145 -> rx len 172
12:02:43.145 -> payload: index 1404, total 65536, avail 172/172
12:02:43.178 -> Parsed 172 - remaining 0
12:02:43.178 -> rx len 740
12:02:43.178 -> payload: index 1576, total 65536, avail 740/740
12:02:43.178 -> Parsed 740 - remaining 0
12:02:43.211 -> rx len 536
12:02:43.211 -> payload: index 2316, total 65536, avail 536/536
12:02:43.211 -> Parsed 536 - remaining 0
12:02:43.211 -> rx len 1072
12:02:43.211 -> payload: index 2852, total 65536, avail 1072/1072
12:02:43.211 -> Parsed 1072 - remaining 0
12:02:43.211 -> rx len 1072
12:02:43.211 -> payload: index 3924, total 65536, avail 1072/1072
12:02:43.211 -> Parsed 1072 - remaining 0
12:02:43.244 -> rx len 1072
12:02:43.244 -> payload: index 4996, total 65536, avail 1072/1072
12:02:43.244 -> Parsed 1072 - remaining 0
12:02:43.244 -> rx len 1072
12:02:43.244 -> payload: index 6068, total 65536, avail 1072/1072
12:02:43.244 -> Parsed 1072 - remaining 0
12:02:43.244 -> rx len 536
12:02:43.244 -> payload: index 7140, total 65536, avail 536/536
12:02:43.244 -> Parsed 536 - remaining 0
12:02:43.244 -> rx len 1440
12:02:43.244 -> payload: index 7676, total 65536, avail 1440/1440
12:02:43.244 -> Parsed 1440 - remaining 0
12:02:43.244 -> rx len 704
12:02:43.244 -> payload: index 9116, total 65536, avail 704/704
12:02:43.277 -> Parsed 704 - remaining 0
12:02:43.277 -> rx len 1072
12:02:43.277 -> payload: index 9820, total 65536, avail 1072/1072
12:02:43.277 -> Parsed 1072 - remaining 0
12:02:43.277 -> rx len 1072
12:02:43.277 -> payload: index 10892, total 65536, avail 1072/1072
12:02:43.277 -> Parsed 1072 - remaining 0
12:02:43.277 -> rx len 536
12:02:43.277 -> payload: index 11964, total 65536, avail 536/536
12:02:43.277 -> Parsed 536 - remaining 0
12:02:43.277 -> rx len 1440
12:02:43.277 -> payload: index 12500, total 65536, avail 1440/1440
12:02:43.310 -> Parsed 1440 - remaining 0
12:02:43.310 -> rx len 704
12:02:43.310 -> payload: index 13940, total 65536, avail 704/704
12:02:43.310 -> Parsed 704 - remaining 0
12:02:43.310 -> rx len 1072
12:02:43.310 -> payload: index 14644, total 65536, avail 1072/1072
12:02:43.310 -> Parsed 1072 - remaining 0
12:02:43.310 -> rx len 1072
12:02:43.310 -> payload: index 15716, total 65536, avail 1072/1072
12:02:43.310 -> Parsed 1072 - remaining 0
12:02:43.310 -> rx len 1072
12:02:43.310 -> payload: index 16788, total 65536, avail 1072/1072
12:02:43.310 -> Parsed 1072 - remaining 0
12:02:43.310 -> rx len 1072
12:02:43.343 -> payload: index 17860, total 65536, avail 1072/1072
12:02:43.343 -> Parsed 1072 - remaining 0
12:02:43.343 -> rx len 1072
12:02:43.343 -> payload: index 18932, total 65536, avail 1072/1072
12:02:43.343 -> Parsed 1072 - remaining 0
12:02:43.343 -> rx len 1072
12:02:43.343 -> payload: index 20004, total 65536, avail 1072/1072
12:02:43.343 -> Parsed 1072 - remaining 0
12:02:43.343 -> rx len 1072
12:02:43.343 -> payload: index 21076, total 65536, avail 1072/1072
12:02:43.343 -> Parsed 1072 - remaining 0
12:02:43.343 -> rx len 1072
12:02:43.343 -> payload: index 22148, total 65536, avail 1072/1072
12:02:43.377 -> Parsed 1072 - remaining 0
12:02:43.377 -> rx len 1072
12:02:43.377 -> payload: index 23220, total 65536, avail 1072/1072
12:02:43.377 -> Parsed 1072 - remaining 0
12:02:43.377 -> rx len 1072
12:02:43.377 -> payload: index 24292, total 65536, avail 1072/1072
12:02:43.377 -> Parsed 1072 - remaining 0
--- network disconnect here ---
12:02:58.118 -> Alloc (l:2)
12:02:58.118 -> tx 2/2 (c0)
12:03:13.369 -> Disconnecting, server exceeded keepalive
12:03:13.668 -> clearing queue (clear session: false)
12:03:13.668 -> ---> ota failed!!!
--- network reconnect here ---
12:03:28.684 -> Alloc (l:49)
12:03:28.684 -> tx 49/49 (10)
12:03:28.750 -> rx len 4
12:03:28.750 -> payload: index 25364, total 65536, avail 4/4
12:03:28.750 -> Disconnecting, expected CONNACK - protocol error
12:03:28.750 -> clearing queue (clear session: false)
```
</details>

> when reconnecting, the parser is still parsing the big message whereas it should parse connack

Looks like that was the case.
I've managed to fix it by reseting **_parse** to **_fixedHeader** before connecting to MQTT:
```cpp
void Parser::reset() {
  _parse = _fixedHeader;
}
```

```diff
    case State::connectingTcp2:
      if (_transport->connected()) {
++      _parser.reset();
        _lastClientActivity = _lastServerActivity = millis();
        _state = State::connectingMqtt;
      }
      break;
```

> It basically comes down to a scenario where receiving a message takes longer then the keepalive time.

I did run some tests like blocking connection in the middle of receiving and unblocking it right after PINGREQ. Packet receiving resumed without any problems. 

<details>
  <summary>
    debug
  </summary>

```
--- skipped rx debug ---
12:56:11.542 -> rx len 1440
12:56:11.542 -> payload: index 34298, total 65536, avail 1440/1440
12:56:11.542 -> Parsed 1440 - remaining 0
12:56:11.542 -> rx len 168
12:56:11.542 -> payload: index 35738, total 65536, avail 168/168
12:56:11.542 -> Parsed 168 - remaining 0
--- network disconnected ---
12:56:26.136 -> Alloc (l:2)
12:56:26.136 -> tx 2/2 (c0)
--- network reconnect ---
12:56:27.429 -> rx len 1072
12:56:27.429 -> payload: index 35906, total 65536, avail 1072/1072
12:56:27.429 -> Parsed 1072 - remaining 0
12:56:27.429 -> rx len 1072
12:56:27.429 -> payload: index 36978, total 65536, avail 1072/1072
--- skipped rx debug ---
12:56:27.694 -> Parsed 536 - remaining 0
12:56:27.694 -> rx len 1224
12:56:27.694 -> payload: index 64314, total 65536, avail 1224/1224
12:56:27.694 -> Alloc (l:4)
12:56:27.694 -> Alloc (l:34)
12:56:27.694 -> Parsed 1222 - remaining 2
12:56:27.694 -> Packet type: 0xd0 <--- ping resp ok ?
12:56:27.694 -> Remaining length: 0
12:56:27.694 -> Parsed 2 - remaining 0
```

![image](https://user-images.githubusercontent.com/2133936/190892478-dfdb5c10-4a54-4dc7-ad17-4fc6e1c498b5.png)

</details>

<details>
  <summary>
    offtopic
  </summary>
Docs says that clean session flags defaults to true, but it actually false.
</details>

**Comment by bertmelis** [comment]
That was indeed the solution I had in mind. Maybe also do a reset on the "incomingPacket" in the reset method, just to be on the safe side.

**Comment by bertmelis** [comment]
Should be solved by #35 .

Feel free to reopen if it hasn't.
